envoy/stats/allocator.h

```diff
@@ -55,6 +55,18 @@ class Allocator {
    */
   virtual TextReadoutSharedPtr makeTextReadout(StatName name, StatName tag_extracted_name,
                                                const StatNameTagVector& stat_name_tags) PURE;
+
+  /**
+   * @param name the full name of the group.
+   * @param tag_extracted_name the name of the stat with tag-values stripped out.
+   * @param tags the tag values.
+   * @param descriptor the maximum number of entries in the group.
+   * @return CounterSharedPtr a counter.
+   */
+  virtual CounterGroupSharedPtr makeCounterGroup(StatName name, StatName tag_extracted_name,
+                                                 const StatNameTagVector& stat_name_tags,
+                                                 CounterGroupDescriptorSharedPtr descriptor) PURE;
```

Comment:
Thinking about the usage model, I'm wondering where. the descriptors will come from.

I think it's probably OK to be unopinionated at this level. You could have the caller create a descriptor and then make a CounterGroup from that.

However what we want to avoid is having multiple copies of the same group, or you lose the benefit of memory savings.

If you leave this in the hands of the caller, I'm wondering whether (eg) Quic will have an appropriate context that is instantiated once per process, in which these descriptors can be explicitly instantiated and referenced. That would in some sense be ideal.

However if Quic (or other possible use-cases for this) doesn't have such a context, we might want to delegate the management of the descriptors to the Stats::Allocator. In that case, we could provide Stats::Allocator::makeCounterGroupDescriptor(...) which would take a string-list and return a canonical descriptor for that. That would need to take a lock on a the allocator's descriptor-cache, and hopefully that would not be overly contended.

Maybe this is worth a TODO here in allocator.h? WDYT?

Reply:
My idea was that we'd have a global function like:

```
CounterGroupDescriptorSharedPtr getQuicCounterGroupDescriptor() {
  static CounterGroupDescriptorSharedPtr descriptor = std::make_shared<CounterGroupDescriptor>({"foo", "bar", "baz"});
  return descriptor;
}
```

This way we allocate a single descriptor which is accessible to any place that want to create a counter group with the same set of suffixes. Does that sound plausible?

Reply:
My personal preference would be for that to come out of a context. I think also we might want to use StatName rather than string as the representation for the enum names within the descriptor, so that they could be sent to StatName::join() without taking a symbol table lock.

Doing that, though, means you need to pass a `SymbolTable&` to the descriptor constructor.

And then wouldn't be able to use a static because in tests, symbol-tables would be re-built with every test method, and you wouldn't be able to destruct the symbol-table for one test method while a static held ownership of live StatNames built it.

Reply:
Interesting. I don't see `StatName::join()` in the code, but I did see `SymbolTable::join()`. Is that the method you're referring to? Can you say more about where this method is called from? In particular, with this PR the `CounterGroup::nameSuffix()` currently returns an `absl::string_view`. Would it need to return a `StatName`?